src/graph/detect_cycle.rs

```diff
@@ -0,0 +1,221 @@
+            for (v, _) in self.adjacency_table().get(u).unwrap() {
+                *in_degree.get_mut(v).unwrap() += 1;
+            }
+        }
+        // Initialize queue with vertex having 0 in-degree
+        let mut queue: VecDeque<&String> = in_degree
+            .iter()
+            .filter(|(_, &degree)| degree == 0)
+            .map(|(&k, _)| k)
+            .collect();
+
+        let mut count = 0;
+        while let Some(u) = queue.pop_front() {
+            count += 1;
+            for (v, _) in self.adjacency_table().get(u).unwrap() {
+                in_degree.entry(v).and_modify(|d| {
+                    *d -= 1;
+                    if *d == 0 {
+                        queue.push_back(v);
+                    }
+                });
+            }
+        }
+
+        // If count of processed vertices is not equal to the number of vertices,
+        // the graph has a cycle
+        count != self.adjacency_table().len()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::data_structures::{graph::Graph, DirectedGraph, UndirectedGraph};
+
+    use super::DetectCycle;
+
+    #[test]
+    fn test_detect_cycle_in_undirected_graph() {
+        let mut graph_with_cycle = UndirectedGraph::new();
+
+        graph_with_cycle.add_edge(("a", "b", 1));
+        graph_with_cycle.add_edge(("a", "c", 1));
+        graph_with_cycle.add_edge(("b", "c", 1));
+        graph_with_cycle.add_edge(("b", "d", 1));
+        graph_with_cycle.add_edge(("c", "d", 1));
+
+        assert!(graph_with_cycle.detect_cycle_dfs());
+        assert!(graph_with_cycle.detect_cycle_bfs());
+
+        let mut graph_without_cycle = UndirectedGraph::new();
+
+        graph_without_cycle.add_edge(("a", "b", 1));
+        graph_without_cycle.add_edge(("a", "c", 1));
+        graph_without_cycle.add_edge(("b", "d", 1));
+        graph_without_cycle.add_edge(("c", "e", 1));
+
+        assert!(!graph_without_cycle.detect_cycle_dfs());
+        assert!(!graph_without_cycle.detect_cycle_bfs());
+    }
+
+    #[test]
+    fn test_detect_cycle_in_directed_graph() {
+        let mut graph_with_cycle = DirectedGraph::new();
+
+        graph_with_cycle.add_edge(("b", "a", 1));
+        graph_with_cycle.add_edge(("c", "a", 1));
+        graph_with_cycle.add_edge(("b", "c", 1));
+        graph_with_cycle.add_edge(("c", "d", 1));
+        graph_with_cycle.add_edge(("d", "b", 1));
+
+        assert!(graph_with_cycle.detect_cycle_dfs());
+        assert!(graph_with_cycle.detect_cycle_bfs());
+
+        let mut graph_without_cycle = DirectedGraph::new();
+
+        graph_without_cycle.add_edge(("b", "a", 1));
+        graph_without_cycle.add_edge(("c", "a", 1));
+        graph_without_cycle.add_edge(("b", "c", 1));
+        graph_without_cycle.add_edge(("c", "d", 1));
+        graph_without_cycle.add_edge(("b", "d", 1));
+
+        assert!(!graph_without_cycle.detect_cycle_dfs());
+        assert!(!graph_without_cycle.detect_cycle_bfs());
+    }
+}
```

Comment:
What do you think about expressing the tests in a more _parametrized_ way? By the way, I think I have found two scenarios, in which I get a wrong result - they are currently _commented out_.

```suggestion
#[cfg(test)]
mod test {
    use crate::data_structures::{graph::Graph, DirectedGraph, UndirectedGraph};

    use super::DetectCycle;

    fn get_undirected_single_node_with_loop() -> UndirectedGraph {
        let mut res = UndirectedGraph::new();

        res.add_edge(("a", "a", 1));
        res
    }

    fn get_directed_single_node_with_loop() -> DirectedGraph {
        let mut res = DirectedGraph::new();

        res.add_edge(("a", "a", 1));
        res
    }

    fn get_undirected_two_nodes_connected() -> UndirectedGraph {
        let mut res = UndirectedGraph::new();

        res.add_edge(("a", "b", 1));
        res
    }

    fn get_directed_two_nodes_connected() -> DirectedGraph {
        let mut res = DirectedGraph::new();

        res.add_edge(("a", "b", 1));
        res.add_edge(("b", "a", 1));
        res
    }

    fn get_directed_two_nodes() -> DirectedGraph {
        let mut res = DirectedGraph::new();

        res.add_edge(("a", "b", 1));
        res
    }

    fn get_undirected_triangle() -> UndirectedGraph {
        let mut res = UndirectedGraph::new();

        res.add_edge(("a", "b", 1));
        res.add_edge(("b", "c", 1));
        res.add_edge(("c", "a", 1));
        res
    }

    fn get_directed_triangle() -> DirectedGraph {
        let mut res = DirectedGraph::new();

        res.add_edge(("a", "b", 1));
        res.add_edge(("b", "c", 1));
        res.add_edge(("c", "a", 1));
        res
    }

    fn get_undirected_triangle_with_tail() -> UndirectedGraph {
        let mut res = get_undirected_triangle();

        res.add_edge(("c", "d", 1));
        res.add_edge(("d", "e", 1));
        res.add_edge(("e", "f", 1));
        res.add_edge(("g", "h", 1));
        res
    }

    fn get_directed_triangle_with_tail() -> DirectedGraph {
        let mut res = get_directed_triangle();

        res.add_edge(("c", "d", 1));
        res.add_edge(("d", "e", 1));
        res.add_edge(("e", "f", 1));
        res.add_edge(("g", "h", 1));
        res
    }

    fn get_undirected_graph_with_cycle() -> UndirectedGraph {
        let mut res = UndirectedGraph::new();

        res.add_edge(("a", "b", 1));
        res.add_edge(("a", "c", 1));
        res.add_edge(("b", "c", 1));
        res.add_edge(("b", "d", 1));
        res.add_edge(("c", "d", 1));
        res
    }

    fn get_undirected_graph_without_cycle() -> UndirectedGraph {
        let mut res = UndirectedGraph::new();

        res.add_edge(("a", "b", 1));
        res.add_edge(("a", "c", 1));
        res.add_edge(("b", "d", 1));
        res.add_edge(("c", "e", 1));
        res
    }

    fn get_directed_graph_with_cycle() -> DirectedGraph {
        let mut res = DirectedGraph::new();

        res.add_edge(("b", "a", 1));
        res.add_edge(("c", "a", 1));
        res.add_edge(("b", "c", 1));
        res.add_edge(("c", "d", 1));
        res.add_edge(("d", "b", 1));
        res
    }

    fn get_directed_graph_without_cycle() -> DirectedGraph {
        let mut res = DirectedGraph::new();

        res.add_edge(("b", "a", 1));
        res.add_edge(("c", "a", 1));
        res.add_edge(("b", "c", 1));
        res.add_edge(("c", "d", 1));
        res.add_edge(("b", "d", 1));
        res
    }

    macro_rules! test_detect_cycle {
        ($($name:ident: $test_case:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let (graph, has_cycle) = $test_case;
                    assert_eq!(graph.detect_cycle_dfs(), has_cycle);
                    assert_eq!(graph.detect_cycle_bfs(), has_cycle);
                }
            )*
        };
    }

    test_detect_cycle! {
        undirected_empty: (UndirectedGraph::new(), false),
        directed_empty: (DirectedGraph::new(), false),
        // undirected_single_node_with_loop: (get_undirected_single_node_with_loop(), true),
        directed_single_node_with_loop: (get_directed_single_node_with_loop(), true),
        // undirected_two_nodes_connected: (get_undirected_two_nodes_connected(), true),
        directed_two_nodes_connected: (get_directed_two_nodes_connected(), true),
        directed_two_nodes: (get_directed_two_nodes(), false),
        undirected_triangle: (get_undirected_triangle(), true),
        undirected_triangle_with_tail: (get_undirected_triangle_with_tail(), true),
        directed_triangle: (get_directed_triangle(), true),
        directed_triangle_with_tail: (get_directed_triangle_with_tail(), true),
        undirected_graph_with_cycle: (get_undirected_graph_with_cycle(), true),
        undirected_graph_without_cycle: (get_undirected_graph_without_cycle(), false),
        directed_graph_with_cycle: (get_directed_graph_with_cycle(), true),
        directed_graph_without_cycle: (get_directed_graph_without_cycle(), false),
    }
}
```

Reply:
Thank you for pointing out the missing logic for `undirected_graph` logic, through the `undirected_single_node_with_loop` test case. For the `undirected_two_nodes_connected`  test I think there is no cycle. 

```rust
res.add_edge(("a", "b", 1));
```
a -- b

Reply:
You are right. Sorry for the confusion.